Amazon.Ion.ObjectMapper/IonSerializer.cs

```diff
@@ -286,84 +259,58 @@ public object Deserialize(IIonReader reader, Type type)
 
         public object Deserialize(IIonReader reader, Type type, IonType ionType)
         {
-            if (ionType == IonType.None || ionType == IonType.Null)
-            {
-                return new IonNullSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Bool)
-            {
-                return new IonBooleanSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Int)
+            if (reader.CurrentDepth >= this.options.MaxDepth)
             {
-                if (reader.GetTypeAnnotations().Any(s => s.Equals(IonLongSerializer.ANNOTATION)))
-                {
-                    return new IonLongSerializer().Deserialize(reader);
-                }
-                return new IonIntSerializer().Deserialize(reader);
+                return null;
             }
 
-            if (ionType == IonType.Float)
+            switch (ionType)
             {
-                if (reader.GetTypeAnnotations().Any(s => s.Equals(IonFloatSerializer.ANNOTATION)))
-                {
-                    return new IonFloatSerializer().Deserialize(reader);
-                }
-                return new IonDoubleSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Decimal)
-            {
-                if (reader.GetTypeAnnotations().Any(s => s.Equals(IonDecimalSerializer.ANNOTATION)))
-                {
-                    return new IonDecimalSerializer().Deserialize(reader);
-                }
-                return new IonBigDecimalSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Blob) 
-            {
-                if (reader.GetTypeAnnotations().Any(s => s.Equals(IonGuidSerializer.ANNOTATION))
-                    || type.IsAssignableTo(typeof(Guid)))
-                {
-                    return new IonGuidSerializer(options).Deserialize(reader);
-                }
-                return new IonByteArraySerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.String) 
-            {
-                return new IonStringSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Symbol) 
-            {
-                return new IonSymbolSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Timestamp) 
-            {
-                return new IonDateTimeSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Clob) 
-            {
-                return new IonClobSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.List) 
-            {
-                return NewIonListSerializer(type).Deserialize(reader);
-            }
-
-            if (ionType == IonType.Struct) 
-            {
-                return new IonObjectSerializer(this, options, type).Deserialize(reader);
+                case IonType.None:

```

Comment:
I don't think switch statements are readable when they get this big, especially considering the nested ifs within them. I think you should put this back the way it was.

Reply:
Reverted back to old way.